include/net/net_tls.h

```diff
@@ -0,0 +1,114 @@
+/** @file
+ * @brief Network TLS definitions
+ *
+ * An API for applications to configure TLS credentials.
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __NET_TLS_H
+#define __NET_TLS_H
+
+/**
+ * @brief Network TLS management
+ * @defgroup net_tls Network TLS management
+ * @ingroup networking
+ * @{
+ */
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** TLS credential type */
+enum net_tls_credential_type {
+	NET_TLS_CREDENTIAL_UNUSED,
+	NET_TLS_CREDENTIAL_CA_CERTIFICATE,
+	NET_TLS_CREDENTIAL_PSK,
+	NET_TLS_CREDENTIAL_PSK_ID
+};
+
+/** Secure tag, a reference for TLS credential */
```

Comment:
How is this supposed to be used? Why exactly it says "TLS credential"? What about "TLS options" and "TLS settings"? How tags with parameters are supposed to be used?

Reply:
By credential we represent an entity that can be used for authentication in TLS  - PSK/PSK_id pair, CA certificate or public certificate/private key pair (not present in this PR). Those are registered in the system with a TAG, that will be used for referencing the credential. A single TAG can reference multiple credentials of different type (for example PSK and PSK_id).

This PR provides the simplest possible system for registering credentials - add, remove, read, not necessiraly thread-safe as you've pointed out. I guess more advanced mechanisms can be added in the feature if needed (like involing filesystem, or some other, secure, storage).

Reply:
> By credential we represent an entity that can be used for authentication in TLS

Ok, so that's conventional definition of credential, thanks. But what about other TLS parameters a user may want to configure - starting from allowed TLS versions for a particular connection, to specific cihersuites/curves/etc. allowed/disallowed?

Reply:
>  But what about other TLS parameters a user may want to configure - starting from allowed TLS versions for a particular connection, to specific cihersuites/curves/etc. allowed/disallowed?

Currently TSL version and ciphersuite configuration is done system-wide, with Kconfig.

> allowed TLS versions for a particular connection

The idea is that this could be selected during socket creation (`IPPROTO_TLS_1_2`, `IPPROTO_TLS_1_1` etc.). Yet, limiting the TLS version is not implemented, and we're supporting all TLS versions selected by Kconfig.

> to specific cihersuites/curves/etc. allowed/disallowed

Changing this in a runtime is not implemented, mebdTLS will use all ciphersuites that were configured statically. Adding an option to configure allowed ciphersuites per TLS context shouldn't be a problem though, as mbedTLS provides convenient API to do that (`mbedtls_ssl_conf_ciphersuites`).

Reply:
> Adding an option to configure allowed ciphersuites per TLS context 

So, my suggestion is to add an example implementation of such an option. Because otherwise it's about the same situation as with Zstream - I avoided wrapping too many mbed calls in ztls_ calls. But here we heard socket options, we need socket options - but there're barely any options defined to show how it'll actually work.

Reply:
> Adding an option to configure allowed ciphersuites per TLS context shouldn't be a problem though, as mbedTLS provides convenient API to do that (mbedtls_ssl_conf_ciphersuites).

For my own reference, there's also:

~~~
 * All lists are bitfields, built by ORing flags from MBEDTLS_X509_ID_FLAG().
 */
typedef struct
{
    uint32_t allowed_mds;       /**< MDs for signatures         */
    uint32_t allowed_pks;       /**< PK algs for signatures     */
    uint32_t allowed_curves;    /**< Elliptic curves for ECDSA  */
    uint32_t rsa_min_bitlen;    /**< Minimum size for RSA keys  */
}
mbedtls_x509_crt_profile;
~~~


So, mbedTLS allows for quite detailed runtime configuration of TLS params.